### PR TITLE
feat: optional checksum algorithm for upload

### PR DESCRIPTION
**Pull request by wuuxigh**
#### Description of changes
- Added optional `checksumAlgorithm` option for `uploadData` API to enable CRC-32 checksum calculation.
- CRC-32 now works within react native environment upon enablement with this new option.
- Upload cache key now includes a `UploadDataOptions` CRC-32 hash to avoid collision.
- Remove `size` from listParts response

#### Description of how you validated changes

- Updated unit tests utilizing the `checksumAlgorithm` option for CRC-32.
- Local run of react native e2e test for 0.70.0.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

#### Checklist for repo maintainers

- [ ] Verify E2E tests for existing workflows are working as expected or add E2E tests for newly added workflows
- [ ] New source file paths included in this PR have been added to CODEOWNERS, if appropriate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/providers/s3/utils/crc32.native.ts [hunk 1]
```diff
@@ -1,11 +1,72 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-import { CRC32Checksum } from './crc32';
+import crc32 from 'crc-32';
+
+import { hexToArrayBuffer, hexToBase64 } from './hexUtils';
+import { readFile } from './readFile';
+
+const CHUNK_SIZE = 1024 * 1024; // 1MB chunks
```
Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/utils/crc32.native.ts:9]: I think this chunk size adds up to the complexity a lot. According to current implementation, the `calculateContentCRC32` function is only called to handle 1 upload part. It only need to handle data bigger than 5 MB when uploading the file size greater than 50GB(max 10k parts). On RN, upload this size of file is even less of a concern. 

I think we can use calculate the CRC32 checksum input content holisticly.

File: packages/storage/src/providers/s3/utils/readFile.ts
Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/utils/readFile.ts:1]: Should this file be `readFile.native.ts`? The `readFile.ts` that's for browser should be moved out of the `md5.ts` as well.

File: packages/storage/src/providers/s3/utils/crc32.native.ts (hunk 1 above)
Comment (wuuxigh) [reply packages/storage/src/providers/s3/utils/crc32.native.ts:9]: The `calculateContentCRC32` function should stay flexible enough to support future development.

File: packages/storage/src/providers/s3/utils/readFile.ts
Comment (wuuxigh) [reply packages/storage/src/providers/s3/utils/readFile.ts:1]: Addressed in new commit.

File: packages/storage/src/providers/s3/utils/getCombinedCrc32.native.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+		crc32List.push(checksum);
+	}
+
+	// Combine all Uint8Arrays into a single Uint8Array
+	const combinedArray = new Uint8Array(totalLength);
+	let offset = 0;
+	for (const crc32Hash of crc32List) {
+		combinedArray.set(crc32Hash, offset);
+		offset += crc32Hash.length;
+	}
```
Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/utils/getCombinedCrc32.native.ts:43]: Thanks @wuuxigh for handling the RN issue with handling creating blob from array buffer!

**Review by AllanZhengYP — APPROVED** [review]
Looks good to me! Thank you for applying the changes!

File: packages/storage/__tests__/providers/s3/utils/getCombinedCrc32.native.test.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+	TextDecoder as TextDecoderPolyfill,
+	TextEncoder as TextEncoderPolyfill,
+} from 'node:util';
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/utils/getCombinedCrc32.native.test.ts:7]: nit: why bring this specifically from `node:util`? We should just do this similar to our core [like this](https://github.com/aws-amplify/amplify-js/blob/main/packages/core/__tests__/singleton/Singleton.test.ts#L1). Any reason not to? 

can be followed up

File: tsconfig.json [hunk 4]
```diff
@@ -25,5 +25,5 @@
 		"module": "es2020",
 		"types": ["node", "jest"]
 	},
-	"exclude": ["node_modules", "dist", ".eslintrc.js", "scripts"]
+	"exclude": ["node_modules", "dist", ".eslintrc.js"]
```
Comment (ashika112) [inline tsconfig.json:28]: why was this changed?

Comment (wuuxigh) [reply tsconfig.json:28]: It fixes a broken pre-commit hook with ESLint.

Comment (AllanZhengYP) [reply tsconfig.json:28]: @ashika112 This fixes an issue with my previous change to dts-bundler script. We do need to enable ts checking and eslint for the dts-bundler script. So we remove it from `exclude`

**Review by ashika112 — APPROVED** [review]
